core/database_api.php

```diff
@@ -332,9 +334,10 @@ function db_query_bound() {
  * @param array   $p_arr_parms Array of parameters matching $p_query.
  * @param integer $p_limit     Number of results to return.
  * @param integer $p_offset    Offset query results for paging.
+ * @param boolean $p_pop_param Set to false to leave the parameters on the stack
  * @return IteratorAggregate|boolean adodb result set or false if the query failed.
  */
-function db_query( $p_query, array $p_arr_parms = null, $p_limit = -1, $p_offset = -1 ) {
+function db_query( $p_query, array $p_arr_parms = null, $p_limit = -1, $p_offset = -1, $p_pop_param = true ) {
```

Comment:
OK, not sure if its the same reasons you had in mind:
In the situation of a nested query call, the inner query always pops the params.
So, even if the inner query dont use db_param in any way, we should also call db_param_push to keep it safe.

This doesn't sound quite right.
In this PR you are using db_param_push() in every api function that uses db_param().
But what about other functions with queriess that dont use db_params()? those are a failure point too... unless you modify the db_query() call to include this parameter to false.

I suggest to check inside db_query if the query has params. If not, then do not pop().